scripts/build.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python
+
+# Copyright 2020 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Use this script to configure and build the firebase cpp sdk
+If you see any issues in configure when run multiple times, 
+try deleting your build directory and use this script after that.
+
+Usage:
+# Build all targets with default options
+python build.py
+
+# Build all targets with default options and also build unit tests
+python build.py --build_tests
+
+# Build all targets without using vcpkg
+# Assuming cpp dependencies are installed and accessible to the build system
+python build.py --no_vcpkg
+
+"""
+
+import subprocess
+import os
+import sys
+import multiprocessing
+import argparse
+
+
+def run_command(cmd, cwd=None, as_root_user=False, collect_output=False):
+    """Run a command.
+
+    Args:
+        cmd (:obj:`list` of :obj:`str`): Command to run as a list object.
+                                         Eg: ['ls', '-l'].
+        cwd (str): Directory to execute the command from.
+        as_root_user (bool): Run command as root user with admin priveleges (supported on mac and linux).
+        collect_output (bool): Get output from the command. The string returned from this function
+                               call will have the output from command execution.
+    Raises:
+        (IOError): If command errored out.
+    Returns:
+        (str|None): If collect_output is provided, a string is returned
+                    None if output is not requested
+    """
+    if as_root_user and (is_mac_os() or is_linux_os()):
+        cmd.insert(0, 'sudo')
+
+    cmd_string = ' '.join(cmd)
+    print('Running cmd: {0}\n'.format(cmd_string))
+    try:
+        if collect_output:
+            output = subprocess.check_output(cmd, cwd=cwd)
+            return output.decode('utf-8').strip()
+        else:
+            subprocess.call(cmd, cwd=cwd)
+    except subprocess.CalledProcessError:
+        raise IOError('Error executing {0}'.format(cmd_string))
+
+
+def get_repo_root_dir():
+    """ Get root directory of current git repo.
+
+    This function takes the guesswork out when executing commands
+    or locating files relative to root of the repo.
+
+    Raises:
+        IOError: Current directory is not in a git repo.
+
+    Returns:
+        (str): Root directory of git repo.
+    """
+    return run_command(['git', 'rev-parse', '--show-toplevel'], collect_output=True)
+
+
+REPO_ROOT_DIR = get_repo_root_dir()
```

Comment:
Yeah, I wanted to cache the value as its used in multiple places. (wanted to use a `lru_cache` from Python3 but that broke Python2 and writing a cache function just for this would have made it look complicated).
Re: why use 'git rev-parse', I thought it will be nice to allow execution of this python script from any subdirectory in the repo and it just works. I could assume users are always at the root but when I am debugging, I am sometimes in the directory containing python script and it will be convenient if I could just run it.